전남대 BE_이현기 3주차 과제 (1단계)

geist9110 opened the pull request:
# 진행상황
- 각 repository별 유닛 테스트를 작성했습니다.
- Product controller에 대해 유닛 테스트를 작성했습니다. 추후에 나머지 Controller와 Service에 대해서도 유닛 테스트를 작성하고 E2E 테스트를 작성하겠습니다.
- 도메인 객체와 Repository 객체들을 JPA를 사용하도록 변경했습니다.

# 궁금한점
- 현재 테스트 코드를 임의로 작성했는데, 작성을 하면서 void값을 리턴하는 메서드들에 대해서는 어떻게 테스트를 해야할지 잘 모르겠습니다. 특히 Controller 영역에서 service레이어에서 return하는 값이 없는 `addProduct`와 같은 메서드들은 어떻게 테스트를 진행해야 좋은 테스트가 될까요?

geist9110 committed:
test: Add ProductControllerTest

geist9110 committed:
docs: Write list of implementation features for step1

geist9110 committed:
chore: Add assertJ dependency for test

geist9110 committed:
test: Add productRepository test

test: Add deleteProduct test

test: Add updateProductTest and Add setup method

Each method run setup because another test case effect the other test

test: Add getAllProductsTest

test: Add addProduct test

geist9110 committed:
test: Add memberRepository test

test: Add findMemberByEmail test

test: Add eixstMemberByEmail test

test: Add addMember test

geist9110 committed:
test: Add wishlistRepository test

test: Add existWishlist test

test: Add deleteWishlist test

test: Add addWishlist test

test: Add getAllWishlists test

geist9110 committed:
test: Add existProduct test

geist9110 committed:
chore: Add JPA dependency and Remove schema.sql

geist9110 committed:
refactor: Modify entities using @Entity annotation

geist9110 committed:
test: Add setup method for not pollution repository

geist9110 committed:
refactor: Modify Product class and refactor ProductRepository class

Product class use class type declaration from record
ProductRepository extend JpaRepository interface
Changed method name depend on ProductRepository
Add Debug properties for transaction and jdbcTemplate

geist9110 committed:
refactor: Refactor Member class and refactor MemberRepository class

Member class use type declaration from record
MemberRepository extend JpaRepository interface
findMemberByEmailTest code modified

geist9110 committed:
refactor: Refactor Wishlist class and refactor WishlistRepository class

Wishlist class use type declaration from record
Add id value in Wishlist class
WishlistRepository extend JpaRepository interface

geist9110 committed:
fix: Fix wishlist id column auto increment

soominsohn commented on src/main/java/gift/member/Member.java:
```diff
@@ -1,13 +1,55 @@
 import jakarta.validation.constraints.Email;
 
-public record Member(
+@Entity
```
이번 미션에서 JPA를 사용하셨군요!
JPA의 특징과 장,단점에 대해 정리해 보시면 좋습니다.

geist9110 replied:
JPA는 ORM이 단순한 쿼리들을 대신해서 처리해주기에 편리하고 좋고, DB에 의존하지 않기에 DB가 변경되어도 실제 코드에는 변경사항이 없다는 것이 장점입니다. 하지만, 조금만 쿼리가 복잡해지면 성능이 떨어지거나, 아예 쿼리문 자체를 직접 작성해야하는 단점도 있습니다.

soominsohn commented on src/test/java/gift/product/ProductControllerTest.java:
```diff
@@ -0,0 +1,159 @@
+                HttpStatus.BAD_REQUEST
+            ),
+            Arguments.of(
+                new ProductDTO("Special😀", 200, "SpecialCharacter-image-url"),
```
😃  👍

soominsohn commented on src/test/java/gift/product/ProductRepositoryTest.java:
```diff
@@ -0,0 +1,117 @@
+    @BeforeEach
+    void setup() {
+        jdbcTemplate.execute("DELETE FROM product");
+        jdbcTemplate.execute("ALTER TABLE product ALTER COLUMN id RESTART WITH 1");
```
✏️ DELETE으로 테이블을 삭제하고 생성하면 id값이 다시 1부터 생성되지 않죠?
`TRUNCATE`에 대해서도 학습해볼까요?

geist9110 replied:
H2 DB는 TRUNCATE 만으로 id값이 초기화되질 않아 뒤에 `RESTART IDENTITY`를 추가해서 auto_increment 값을 초기화 시켜줬습니다.

soominsohn commented on src/main/java/gift/product/Product.java:
```diff
@@ -1,5 +1,80 @@
+                   && this.imageUrl.equals(product.imageUrl);
+        }
+        return false;
+    }
```
equals메서드를 재정의 하셨군요!
자바 객체에서는 이펙티브자바 item 11. equals를 재정의하려거든 hashCode도 재정의하라 라는 말이 있어요.
https://jake-seo-dev.tistory.com/28

그러나 자바 객체가 JPA entity로 쓰인다면 hashCode를 재정의 해야할까요? 지금 꼭 참고하실 필요는 없고 아래글 보시고 생각을 정리해보면 좋습니다.
https://blog.yevgnenll.me/posts/jpa-entity-eqauls-and-hashcode-equality

geist9110 replied:
JPA entity를 쓰더라도 hashCode는 같이 재정의 해주는게 좋을거 같습니다.hashCode는 Map과 같은 자주쓰는 자료구조의 성능의 차이를 낼 수 있는 중요한 코드입니다. 만일, 다른 사람이 Map 자료구조를 사용한다면 성능면에서 많은 손실을 가져오고 디버깅도 쉽지않기 때문에 재정의가 필요할거 같습니다.

soominsohn commented on src/main/resources/application.properties:
```diff
@@ -5,10 +5,18 @@ spring.thymeleaf.suffix=.html
 spring.h2.console.enabled=true
 spring.h2.console.path=/h2-console
-spring.datasource.url=jdbc:h2:mem:spring-gift
+spring.datasource.url=jdbc:h2:mem:spring-gift;MODE=MYSQL;DB_CLOSE_DELAY=-1
```
✏️ 테스트 환경에서도 서비스의 데이터베이스를 사용하고 있어요.
테스트환경의 설정을 따로 해주는 방법이 있을지 찾아보고 적용해보아요 😄

soominsohn commented on src/main/java/gift/member/Member.java:
```diff
@@ -1,13 +1,55 @@
 
-public record Member(
+@Entity
+public class Member {
```
✏️ 또 이건 이전 피드백으로 들었어야했는데,
MemberDTO가 있음에도 불구하고 MemberController의 RequestBody로 Member를 받고 있어요. 적절히 수정해볼까요? 😃

soominsohn commented on src/main/java/gift/product/ProductService.java:
```diff
@@ -28,7 +28,7 @@ public void addProduct(ProductDTO productDTO) {
+        productRepository.save(
             new Product(
                 id,
                 productDTO.name(),
```
product를 update하려고 하시는군요!
다음과 같이 database에 존재하지 않는 id로 update요청을 날리면 어떻게 될까요?
```
PATCH http://localhost:8080/api/products/1234
Content-Type: application/json

{
  "name": "fsf",
  "price": 1111,
  "imageUrl": "fsad"
}
```

geist9110 replied:
patch는 존재하는 값에 한정해서 업데이트 쿼리를 보내야하는데, 존재하는지에 대한 검증 로직이 빠져 바로 저장되는 문제가 있습니다.

soominsohn commented on src/main/java/gift/wishlist/Wishlist.java:
```diff
@@ -1,5 +1,63 @@
+
+    public void setId(long id) {
+        this.id = id;
+    }
```
✏️ setter가 반복적으로 사용되고 있어요.
`객체 지향 생활 체조 원칙 규칙 9: getter/setter/프로퍼티를 쓰지 않는다.` 에 대해 학습해보고 적용해 볼까요?
객체 지향 생활 체조 원칙 규칙 내용은 모두 학습해보면 좋습니다.

geist9110 replied:
말씀해주신대로 사용하지 않은 getter와 setter를 최대한 제거해봤는데, 특정 getter의 경우 제거할 방법이 도저히 떠오르지가 않습니다... 메시지를 전달하는 방법으로 시도를 해봤으나 실패했습니다... 특히 DB에서 `findBy`를 통해 값을 찾으려고 하는 로직이 하나라도 있으면 그 값의 getter를 도저히 제거할 방법을 모르겠습니다.

soominsohn replied:
네 맞습니다!
값을 전달하기 위한 getter는 사용해도 좋아요. 
그러나 값을 전달하기 위한 목적이 아닌 다른 목적으로 사용하게 된다면 객체지향적이지 않은 코드가 될 수도 있답니다.
예) getter로 가져온 값으로 조건 검사를 하게되는경우 등
앞으로 더 다양한 코드를 작성하다 보면, 어떤 상황에서 사용해야 하고 사용하지 말아야 할지 감을 잡게 될 것입니다.

https://colabear754.tistory.com/173

soominsohn commented on src/main/java/gift/product/Product.java:
```diff
@@ -1,5 +1,80 @@
+    long id;
+
+    @Column(nullable = false)
+    @Length(min = 1, max = 15)
```
Entity에서도 유효성검사를 해주고 계시군요!
이전 단계에서 배운 어노테이션을 사용하는 것 말고 다른방법으로도 검증해 볼 수 있지 않을까요? 개선해보면 좋겠습니다.

geist9110 replied:
잘못 코딩했습니다.. 이 부분도 검증을 위한게 아니라, `VARCHAR(15)`를 생각하고 넣었는데 멘토님이 말씀해주신 것처럼 Validation을 위한 어노테이션이었습니다.. `@Column(length=15)`로 수정하였습니다.

soominsohn commented on src/main/java/gift/wishlist/WishlistService.java:
```diff
@@ -21,24 +21,27 @@ public WishlistService(
+            .stream()
+            .map(e -> productRepository.findById(e.getProductId()).get())
+            .toList();
     }
```
getAllWishlists함수에서 Product list를 반환하고 있어요. 그렇게 짜신 이유가 있나요?
✏️ 또 `spring-data-jpa`의 `findby` 메서드를 사용하셨는데요, return 값은 어떻게 되며 `get()`을 사용하게 된다면 어떤 문제점이 생길 수 있고 어떻게 개선해 볼 수 있을까요?

geist9110 replied:
- Product list를 반환한 이유는 사용자 입장에서 특정 사용자가 위시리스트에 추가한 Product 객체들을 보는걸 원하는게 아닐까라는 생각에서 Product list를 반환했습니다.
- JPA의 findBy는 `Optional`객체를 반환합니다. 이때 findBy를 통해 `null`값이 들어올 수 있는데, 이때 `get()`을 하게 되면 `NullPointException`이 발생합니다. get전에 `isPresent()` 메서드를 필터에 통과시켜 전처리한다면 개선할 수 있을거라 생각합니다.

soominsohn commented on src/main/java/gift/wishlist/Wishlist.java:
```diff
@@ -1,5 +1,63 @@
+public class Wishlist {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
id의  `@GeneratedValue `strategy로 `GenerationType.IDENTITY`를 사용하셨군요.
어떤 strategy들이 또 있을까요? 😄

geist9110 replied:
`SEQUENCE`, `TABLE`, `AUTO` 이렇게 3가지가 더 있습니다. `IDENTITY`나 `SEQUENCE`가 보편적인 방법으로 알고있어서 `IDENTITY`를 사용했습니다.

soominsohn commented on src/test/java/gift/member/MemberRepositoryTest.java:
```diff
@@ -0,0 +1,73 @@
+
+    @BeforeEach
+    void setUp() {
+        memberRepository.deleteAll();
```
테스트 격리를 위한 코드 👍 
다른 `테스트 격리`방법엔 무엇이 있을지 학습해 볼까요?

geist9110 replied:
`Mock`을 사용해 DB를 오염시키지 않고 진행하는 방법과 `@Transactional` 어노테이션을 사용해 각 테스트때 사용했던 오염을 테스트가 끝나고 다시 롤백 시키는 방법이 있습니다.

soominsohn commented on src/test/java/gift/product/ProductRepositoryTest.java:
```diff
@@ -0,0 +1,117 @@
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.jdbc.core.JdbcTemplate;
+
+@DataJpaTest
```
`@DataJpaTest` 적용 👍 
특징에 대해 학습해 보아요 😄

soominsohn commented on src/test/java/gift/product/ProductControllerTest.java:
```diff
@@ -0,0 +1,159 @@
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+@WebMvcTest(ProductController.class)
```
Controller 테스트를 위해 `@WebMvcTest`를 사용하셨군요! `@WebMvcTest`은 어떨 때 사용하나요? 특징에 대해 학습해보아요 ㅎㅎ

geist9110 replied:
`@WebMvcTest`는 컨트롤러 계층을 테스트할때 사용합니다. `@SpringbootTest`를 사용하면 코드내에 있는 모든 Bean들이 만들어지는데 이러면 빌드 시간이 너무 오래 걸려서, Controller를 테스트할 때 필요한 Bean들만 만들어서 테스트를 진행하게 됩니다.

soominsohn replied:
잘 학습해주셨습니다 😄

soominsohn commented on src/test/java/gift/product/ProductControllerTest.java:
```diff
@@ -0,0 +1,159 @@
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private ProductService productService;
```
`@MockBean`활용 👍 
어떻게 동작하는지 정리해두시면 좋습니다 ㅎㅎ

soominsohn commented on src/test/java/gift/product/ProductControllerTest.java:
```diff
@@ -0,0 +1,159 @@
+        );
+    }
+
+    @ParameterizedTest
+    @DisplayName("[Unit] updateProduct test")
+    @MethodSource("updateProductTest")
```
`@ParamterizedTest`활용 👍 
`@MethodSource 활용` 👍

soominsohn commented on src/test/java/gift/product/ProductControllerTest.java:
```diff
@@ -0,0 +1,159 @@
+            .andExpect(content().string(errorMessage));
+    }
+
+    private static Stream<Arguments> updateProductTest() {
```
@MethodSource를 활용하면 Steam을 꼭 사용해야 하나요? 
이건 저도 궁금해서 물어보는겁니다!

geist9110 replied:
꼭 사용해야하는 것은 아닌거 같습니다. 찾아보니 `List`와 `int[][]`, `Object[][]` 이런거도 사용이 가능한거 같습니다.

soominsohn replied:
고맙습니다 ㅎㅎ! 저도 써먹어봐야겠네요

soominsohn commented on src/test/java/gift/product/ProductControllerTest.java:
```diff
@@ -0,0 +1,159 @@
+            .andExpect(jsonPath("$[1].imageUrl").value(product2.getImageUrl()));
+    }
+
+    @ParameterizedTest
+    @DisplayName("[Unit] addProduct test")
+    @MethodSource(value = "addProductTestValues")
+    void addProductTest(ProductDTO productDTO, String errorMessage, HttpStatus httpStatus)
+        throws Exception {
+        mockMvc.perform(post(apiUrl)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(objectMapper.writeValueAsString(productDTO)))
+            .andExpect(status().is(httpStatus.value()))
+            .andExpect(content().string(errorMessage));
+    }
```
> 현재 테스트 코드를 임의로 작성했는데, 작성을 하면서 void값을 리턴하는 메서드들에 대해서는 어떻게 테스트를 해야할지 잘 모르겠습니다. 특히 Controller 영역에서 service레이어에서 return하는 값이 없는 addProduct와 같은 메서드들은 어떻게 테스트를 진행해야 좋은 테스트가 될까요?

주신 질문에 대한 답변입니다!
현기님은 무엇을 테스트 하고싶어서 Controller테스트에서 @WebMvcTest와 mockMvc를 사용하셨을까요?
데이터베이스에 잘 적재되는건 `ProductRepositoryTest`를 이용해 검증하셨어요.
service layer를 mock처리한 Controller테스트는 어떤걸 테스트해 볼 수 있을까요?
`hint: 요청과 응답`

geist9110 replied:
- 우선 API의 경로, 입력값의 검증 절차와 그에 따른 ExceptionHandler들이 제대로 동작하는지, 그리고 반환되는 결과와 Status 코드를 검증하고 싶었습니다.
- Controller 테스트는 앞서 말한 조건들과 Controller레이어에서 Service레이어로 보내주는 값과 Service레이어에서 Controller레이어로 반환하는 값 또한 제대로된 객체를 주고 받았는지 검증할 수 있을거 같습니다.

soominsohn commented on src/main/java/gift/member/Member.java:
```diff
@@ -1,13 +1,55 @@
+    @Column(nullable = false)
+    String password;
+
+    public Member() {
+    }
```
JPA Entity에서 기본생성자가 왜 필요할까요?

geist9110 replied:
Hibernate에서 `ReflectionAPI`를 사용해 구현을 했기 때문입니다. 이때, `ReflectionAPI`는 `클래스명`만을 가지고 값을 조작해 나가기 때문에 기본 생성자로 객체를 생성후, 테이블의 컬럼과 매핑되는 값을 찾아 하나의 완전한 객체로 만듭니다.

soominsohn replied:
오오 ㅎㅎ 잘 학습해주셨군요! 👍

soominsohn commented on src/test/java/gift/wishlist/WishlistRepositoryTest.java:
```diff
@@ -0,0 +1,124 @@
+
+    @BeforeEach
+    void setupMemberAndProduct() {
+        jdbcTemplate.execute("DELETE FROM product");
+        jdbcTemplate.execute("ALTER TABLE product ALTER COLUMN id RESTART WITH 1");
+        jdbcTemplate.execute("DELETE FROM member");
+        jdbcTemplate.execute("DELETE FROM wishlist ");
+        jdbcTemplate.execute("ALTER TABLE wishlist ALTER COLUMN id RESTART WITH 1");
```
이부분도 `TRUNCATE`에 대해 학습해보고 적용해 보시면 좋습니다.

soominsohn reviewed (changes requested):
안녕하세요 현기님. 1단계 잘해주셨네요. 
점점 성장하는게 잘 보이네요. 😄 
학습 열심히 해주신 것 같아 고맙습니다.

개선 및 반영했으면 하는 부분들과 현기님의 생각이 궁금한 부분들 코멘트로 남겨두었으니 확인 부탁드립니다.
양이 많아 **✏️ 표시해둔 코멘트 위주로 반영**해주시고 재리뷰요청주세요.
질문에 대한 답변도 코멘트로 남겨 두었어요.

또 1주차에 작성해주신 selenium테스트는 다른 테스트들과 달리 어플리케이션을 실행시키지 않으면 통과하지  않는데 어떤 문제가 발생할 수 있을지 고민해보면 좋겠습니다.

geist9110 committed:
refactor: Refactor beforeEach delete table to use truncate

geist9110 committed:
refactor: Add properties for separate test db and service db

geist9110 committed:
refactor: Add MemberTokenDTO

MemberController receive MemberDTO
MemberResolver changed to MemberTokenResolver and resolve MemberTokenDTO
JWT receive MemberTokenDTO and return MemberTokenDTO
WishlistController receive MemberTokenDTO

geist9110 committed:
fix: updateProduct save does not exist product id

geist9110 committed:
refactor: Remove setter and getter

but, could not remove all getter in wishlist by JPA findBy method

geist9110 committed:
refactor: Override hashCode and equals method compare their id value

geist9110 committed:
fix: Fix adminPageTest not work

Add product class getter for display adminPage with thymeleaf
Add JWT secret-key in test properties for generate WebConfig bean

geist9110 committed:
refactor: Separate profile in test application.properties

soominsohn reviewed (approved):
안녕하세요 현기님~
1단계 학습 정말 잘해주셨어요. 💯 
많은 내용을 공부하신 것 같네요.
남은 3주차 과제도 파이팅입니다. 👍
